man/man2/reboot.2
.\" @(#)reboot.2 1.1 92/07/30 SMI; from UCB 4.2
.TH REBOOT 2 "21 January 1990"
.SH NAME
reboot \- reboot system or halt processor
.SH SYNOPSIS
.LP
.nf
.ft B
#include <sys/reboot.h>
.ft
.fi
.LP
.nf
.BR "reboot(howto, " [ " bootargs " ] " )"
.ft B
int howto;
char *bootargs;
.ft
.fi
.IX  reboot()  ""  "\fLreboot()\fP \(em halt processor"
.IX  "system operation support"  reboot()  ""  \fLreboot()\fP
.IX  "halt processor"
.IX  "stop processor"
.SH DESCRIPTION
.B reboot(\|)
reboots the system, and is invoked automatically
in the event of unrecoverable system failures.
.I howto
is a mask of options passed to the bootstrap program.
The system call interface permits only
.SB RB_HALT
or
.SB RB_AUTOBOOT
to be passed to the reboot program; the other flags
are used in scripts stored on the console storage media, or used
in manual bootstrap procedures.
When none of these options (for instance
.SM
.BR RB_AUTOBOOT \s0)
is given, the system is rebooted
from file 
.B /vmunix
in the root file system of unit 0
of a disk chosen in a processor specific way.
An automatic consistency check of the disks is then normally performed.
.LP
The bits of
.I howto
are:
.TP 15
.SB RB_HALT
the processor is simply halted; no reboot takes place.
.SB RB_HALT
should be used with caution.
.TP
.SB RB_ASKNAME
Interpreted by the bootstrap program itself, causing it to
inquire as to what file should be booted.  Normally, the system is
booted from the file
.B /vmunix
without asking.
.TP
.SB RB_SINGLE
Normally, the reboot procedure involves an automatic disk consistency
check and then multi-user operations.
.SB RB_SINGLE
prevents the consistency
check, rather simply booting the system with a single-user shell on
the console.
.SB RB_SINGLE
is interpreted by the
.BR init (8)
program in the newly booted system.
.TP
.SB RB_DUMP
A system core dump is performed before rebooting.
.TP
.SB RB_STRING
The optional argument
.I bootargs
is passed to the bootstrap program.
See
.BR boot (8S)
for details.
This option overrides
.SB RB_SINGLE
but the same effect can be achieved by including
.B \-s
as an option in
.IR bootargs .
.LP
Only the super-user may
.B reboot(\|)
a machine.
.SH RETURN VALUES
On success,
.B reboot(\|)
does not return.
On failure,
it returns
\-1
and sets
.B errno
to indicate the error.
.SH ERRORS
.TP 15
.SM EPERM
The caller is not the super-user.
.SH FILES
.PD 0
.TP 20
.B /vmunix
.PD
.SH "SEE ALSO"
.BR panic (8S),
.BR halt (8),
.BR init (8),
.BR intro (8),
.BR reboot (8)


etc/reboot.c
#ifndef lint
static char sccsid[] = "@(#)reboot.c 1.1 92/07/30 SMI"; /* from UCB 5.4 5/26/86 */
#endif

/*
 * Copyright (c) 1980,1986 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1980,1986 Regents of the University of California.\n\
 All rights reserved.\n";
#endif not lint

/*
 * Reboot
 */
#include <stdio.h>
#include <sys/reboot.h>
#include <errno.h>
#include <signal.h>
#include <pwd.h>
#include <strings.h>
#include <syscall.h>
#include <sys/types.h>
#include <sys/label.h>
#include <sys/audit.h>
#include <sys/time.h>
#include <sys/syslog.h>
#include <sys/file.h>

void dingdong();

main(argc, argv)
	int argc;
	char **argv;
{
	int howto = 0;
	char *audit_argv[6];
	register int qflag = 0;
	int needlog = 1;
	char *user, *getlogin();
	struct passwd *pw, *getpwuid();
	extern char *optarg;
	extern int optind, opterr, errno;
	char c;

	openlog("reboot", 0, LOG_AUTH);

	audit_argv[0] = argv[0];
	while ((c = getopt(argc, argv, "qndl")) != EOF)
		switch((char)c) {
		case 'q':
			qflag++;
			break;
		case 'n':
			howto |= RB_NOSYNC;
			break;
		case 'd':
			howto |= RB_DUMP;
			break;
		case 'l':
			needlog = 0;
			break;
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc == 1)
		howto |= RB_STRING;
	else if (argc != 0) 
		usage();

	audit_argv[1] = qflag ? "quick" : "not quick";
	audit_argv[2] = howto & RB_NOSYNC ? "not synced" : "synced";
	audit_argv[3] = howto & RB_DUMP ? "dump" : "no dump";
	audit_argv[4] = access("/fastboot", F_OK) ? "fastboot" : "do fsck";
	audit_argv[5] = howto & RB_STRING ? argv[0] : " ";

	(void)signal(SIGHUP, SIG_IGN);	/* for remote connections */
	if (kill(1, SIGTSTP) == -1) {
		(void)fprintf(stderr, "reboot: can't idle init\n");
		audit_text(AU_MAJPRIV, 1, 1, 6, audit_argv);
		exit(1);
	}

	if (needlog) {
		user = getlogin();
		if (user == NULL && (pw = getpwuid(getuid())) != NULL)
			user = pw->pw_name;
		if (user == NULL)
			user = "root";
		syslog(LOG_CRIT, "rebooted by %s", user);
	}

	sleep(1);
	(void)kill(-1, SIGTERM);	/* one chance to catch it */
	sleep(5);

	if (!qflag)  {
		register unsigned i = 0;
		(void)signal(SIGALRM, dingdong);
		while (kill(-1, SIGKILL) == 0) {
			if (++i == 6) {
				(void)fprintf(stderr,
			 "CAUTION: some process(es) wouldn't die\n");
				errno = 0;
				break;
			}
			(void)alarm(2 * i);
			(void)pause();
		}

		if (errno != 0 && errno != ESRCH) {
			perror("reboot: kill");
			(void)kill(1, SIGHUP);
			audit_text(AU_MAJPRIV, 2, 2, 6, audit_argv);
			exit(1);
		}

		if ((howto & RB_NOSYNC) == 0) {
			markdown();
			sync();
			(void)alarm(5);
			(void)pause();
		}
	}
	syscall(SYS_reboot, howto, argv[0]);
	audit_args(AU_MAJPRIV, 6, audit_argv);
	perror("reboot");
	audit_text(AU_MAJPRIV, 3, 3, 6, audit_argv);
	(void)kill(1, SIGHUP);
	exit(1);
	/* NOTREACHED */
}

void
dingdong()
{
	/* RRRIIINNNGGG RRRIIINNNGGG */
}

#include <utmp.h>
#define SCPYN(a, b)	(void)strncpy(a, b, sizeof (a))
char wtmpf[] = "/var/adm/wtmp";

markdown()
{
	off_t lseek();
	time_t time();
	struct utmp wtmp;
	register int f = open(wtmpf, O_WRONLY, 0);
	if (f >= 0) {
		(void)lseek(f, 0L, L_XTND);
		SCPYN(wtmp.ut_line, "~");
		SCPYN(wtmp.ut_name, "shutdown");
		SCPYN(wtmp.ut_host, "");
		wtmp.ut_time = time((time_t *)0);
		(void)write(f, (char *)&wtmp, sizeof wtmp);
		(void)close(f);
	}
}

usage()
{
	(void)fprintf(stderr,
	    "usage: reboot [ -dnql ] [ boot args ]\n");
	exit(1);
}
